File: backend/app/models/task.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, String, Text, Date, Boolean, DateTime, Enum, SmallInteger, DECIMAL, ForeignKey
from sqlalchemy.orm import relationship

from app.database import Base


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    title = Column(String(200), nullable=False)
    description = Column(Text)

    priority = Column(SmallInteger, nullable=False, default=3, comment="1=urgent .. 5=can wait")
    estimated_duration_minutes = Column(Integer, nullable=False, default=30)

    category = Column(Enum("hobby", "skill", "fitness", "career", "health", "academic", "chore", "other", name="task_category"), nullable=False, default="other")

    due_date = Column(Date)
    is_completed = Column(Boolean, default=False)
    is_recurring = Column(Boolean, default=False)
    recurrence_rule = Column(String(255))
    is_flexible = Column(Boolean, default=True)
    preferred_time_slot = Column(Enum("morning", "afternoon", "evening", "any", name="time_slot"), default="any")

    min_times_per_week = Column(Integer)

    parent_goal_id = Column(Integer, ForeignKey("tasks.id", ondelete="SET NULL"))
    progress = Column(DECIMAL(5, 2), default=0.00)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    user = relationship("User", back_populates="tasks")

File: mobile/lib/models/task.dart
class Task {
  final int id;
  final int userId;
  final String title;
  final String? description;
  final int priority;
  final int estimatedDurationMinutes;
  final String category;
  final DateTime? dueDate;
  final bool isCompleted;
  final bool isFlexible;
  final String preferredTimeSlot;
  final int? minTimesPerWeek;
  final double progress;
  final DateTime createdAt;

  Task({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.priority,
    required this.estimatedDurationMinutes,
    required this.category,
    this.dueDate,
    required this.isCompleted,
    required this.isFlexible,
    required this.preferredTimeSlot,
    this.minTimesPerWeek,
    required this.progress,
    required this.createdAt,
  });

  factory Task.fromJson(Map<String, dynamic> json) => Task(
    id: json['id'] as int,
    userId: json['user_id'] as int,
    title: json['title'] as String,
    description: json['description'] as String?,
    priority: json['priority'] as int,
    estimatedDurationMinutes: json['estimated_duration_minutes'] as int,
    category: json['category'] as String,
    dueDate: json['due_date'] != null ? DateTime.parse(json['due_date'] as String) : null,
    isCompleted: json['is_completed'] as bool? ?? false,
    isFlexible: json['is_flexible'] as bool? ?? true,
    preferredTimeSlot: json['preferred_time_slot'] as String? ?? 'any',
    minTimesPerWeek: json['min_times_per_week'] as int?,
    progress: double.parse(json['progress'].toString()),
    createdAt: DateTime.parse(json['created_at'] as String),
  );

  Map<String, dynamic> toJson() => {
    'title': title,
    'description': description,
    'priority': priority,
    'estimated_duration_minutes': estimatedDurationMinutes,
    'category': category,
    'due_date': dueDate?.toIso8601String().split('T').first,
    'is_flexible': isFlexible,
    'preferred_time_slot': preferredTimeSlot,
    'min_times_per_week': minTimesPerWeek,
  };
}
